chips/msp432/src/spi.rs

```diff
@@ -0,0 +1,476 @@
+//! Serial Peripheral Interface (SPI)
+//!
+//! Implementation of DMA-based SPI master communication for the MSP432.
+//! Both eUSCI_A and eUSCI_B can be used for SPI communication, but the
+//! the registers have a different meaning than in UART/I2C mode.
+//!
+
+use crate::dma;
+use crate::usci::{self, UCSPIxCTLW0, UCSPIxIE, UCSPIxIFG, UCSPIxSTATW};
+use core::cell::Cell;
+use kernel::common::{
+    cells::{OptionalCell, TakeCell},
+    registers::{ReadOnly, ReadWrite},
+    StaticRef,
+};
+use kernel::hil::gpio::Pin;
+use kernel::hil::spi;
+use kernel::ErrorCode;
+
+/// The SPI related registers are identical between the USCI_A and the USCI_B module, but the usage
+/// of the certain bits is different to UART and I2C mode. Thus we simply cast the concerned
+/// registers to a UCSPIx-registers instead of UCAx or UCBx registers. With this trick we can pass
+/// references of USCI_A and USCI_B modules to the Spi-constructor.
+pub trait UsciSpiRef {
+    fn ctlw0(&self) -> &ReadWrite<u16, UCSPIxCTLW0::Register>;
+    fn brw(&self) -> &ReadWrite<u16>;
+    fn statw(&self) -> &ReadWrite<u16, UCSPIxSTATW::Register>;
+    fn rxbuf(&self) -> &ReadOnly<u16>;
+    fn txbuf(&self) -> &ReadWrite<u16>;
+    fn ie(&self) -> &ReadWrite<u16, UCSPIxIE::Register>;
+    fn ifg(&self) -> &ReadWrite<u16, UCSPIxIFG::Register>;
+}
```

Comment:
This seems like a reasonable approach, but could perhaps be explained better. You aren't casting, you are encapsulating. I.e., the two actual bit mappings are different, so you define a common abstraction (`UsciSpiRef`, which should perhaps be `UsciSpiRegisters`), such that the SPI implementation can operate on top of either one.

I would suggest having the Spi type be a type parameter, rather than a `dyn`, though. We've found that this reduces code size and reduces overheads.

Reply:
The underlying type is in fact transmuted in the implementation of this trait.